utils/prof_utils.py

```diff
@@ -0,0 +1,523 @@
+    """
+    Given a profile in which the noise is clipped to 0, finds the components that are clumped together.
+
+    Parameters:
+    -----------
+    profile: list
+        A list of floats describing the profile where the noise has been clipped to zero
+    min_comp_len: float
+        OPTIONAL - Minimum length of a component to be considered real. Measured in bins. Default: 5
+
+    Returns:
+    --------
+    component_dict: dictionary
+        dict["component_x"] contains an array of the component x
+    component_idx: dictionary
+        dict["component_x"] contains an array of indexes of the original profile corresponding to component x
+    """
+    component_dict={}
+    component_idx={}
+    num_components=0
+    for i, val in enumerate(profile):
+        if val!=0.:
+            if profile[i-1]==0 or i==0:
+                num_components+=1
+                comp_key = "component_{}".format(num_components)
+                component_dict[comp_key]=[]
+                component_idx[comp_key]=[]
+            component_dict[comp_key].append(val)
+            component_idx[comp_key].append(i)
+
+    del_comps = []
+    for comp_key in component_dict.keys():
+        if len(component_dict[comp_key]) < min_comp_len:
+            del_comps.append(comp_key)
+    for i in del_comps:
+        del component_dict[i]
+        del component_idx[i]
+
+    return component_dict, component_idx
+
+def find_minima_maxima(profile, ignore_threshold=0.02, min_comp_len=5):
+    """
+    Finds all minima and maxima of the input profile. Assumes that the profile has noise zero-clipped.
+
+    profile: list
+        The profile with noise zero-clipped
+    ignore_threshold: float
+        OPTIONAL -  Default: 0.02
+    min_comp_len: float
+        OPTIONAL - Minimum length of a component to be considered real. Measured in bins. Default: 5
+
+    Returns:
+    --------
+    minima: list
+        A list of floats corresponding to the bin location of the profile minima
+    maxima: list
+        A list of floats corresponding to the bin location of the profile maxima
+    """
+    #If there is more than one component, find each one
+    comp_dict, comp_idx = find_components(profile, min_comp_len)
+
+    maxima=[]
+    minima=[]
+    #loop over each profile component
+    for key in comp_dict.keys():
+        x = np.linspace(0, len(comp_dict[key])-1, len(comp_dict[key]), dtype=int)
+        spline = UnivariateSpline(x, comp_dict[key], s=0.0, k=4)
+        comp_roots = spline.derivative().roots()
+        # These are the roots, we want to split maxima and minima ^^
+        comp_maxima=[]
+        comp_minima=[]
+        for i, root in enumerate(comp_roots):
+            idx = int(root)
+            left = comp_dict[key][idx-1]
+            if left>comp_dict[key][idx]:
+                comp_minima.append(root)
+            else:
+                comp_maxima.append(root)
+        #Turn the root locations into locations on profile, not on component        
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Trailing whitespace](https://app.codacy.com/app/CIRA-Pulsars-and-Transients-Group/vcstools/pullRequest?prid=4645369)